engine/src/main/java/io/camunda/zeebe/streamprocessor/DirectProcessingResultBuilder.java

```diff
@@ -54,6 +68,21 @@ public ProcessingResultBuilder appendRecord(
       final RejectionType rejectionType,
       final String rejectionReason,
       final RecordValue value) {
+
+    final ValueType valueType = typeRegistry.get(value.getClass());
+    if (valueType == null) {
+      // usually happens when the record is not registered at the TypedStreamEnvironment
+      throw new IllegalStateException("Missing value type mapping for record: " + value.getClass());
+    }
+
+    if (value instanceof UnifiedRecordValue unifiedRecordValue) {
+      mutableRecordBatch.appendRecord(
+          key, -1, type, intent, rejectionType, rejectionReason, valueType, unifiedRecordValue);
+    } else {
+      throw new IllegalStateException(
+          String.format("The record value %s is not a UnifiedRecordValue", value));
+    }
```

Comment:
We do a similar thing in the Legacy writer see https://github.com/camunda/zeebe/blob/main/engine/src/main/java/io/camunda/zeebe/streamprocessor/LegacyTypedStreamWriterImpl.java#L85-L89

We could change the interface to accept only UnifiedRecordValue, but this would cause more changes on other interfaces which I didn't wanted to do in this PR.